046-analyse-broker-metrics-before-rolling-kafka.md

```diff
@@ -0,0 +1,93 @@
+
+KafkaAgent currently collects the `kafka.server:type=KafkaServer,name=BrokerState` metric and creates a `kafka-ready` file on disk if the broker state value is greater than `3`(`RUNNING`) and not `127` (`UNKNOWN`). The Kafka readiness check passes if it finds the file on disk.
+
+## Motivation
+
+To address the issues with restarts and log recovery, KafkaRoller needs to be aware of the current broker state.  If a broker is performing log recovery, KafkaRoller should fail the reconciliation instead of force restarting it.
+
+## Proposal
+ 
+The objective of this proposal is to introduce a REST API endpoint within KafkaAgent, which exposes metrics from a Kafka broker. Furthermore, KafkaRoller would be modified so that it can query this endpoint to determine the current state of a broker, thus preventing unnecessary restarts. Additionally, KafkaRoller would allow a human operator to observe whether the log recovery process is making progress or is stuck.
+
+The following implementation details will describe behaviour of the new API in KafkaAgent and how KafkaRoller would interact with it and handle a broker that is in log recovery.
+
+### Implementation details
+
+**KafkaAgent**:
+
+- Add a web server to accept SSL connections on port 8443.
+- Use Jetty for the web server since Kafka already uses it therefore avoids an additional dependency.
+- Reuse the broker's own certificate that would be in the container for TLS authentication. 
+- Configure the server to require TLS client authentication. 
+- Expose an endpoint `v1/recovery-state` that includes the API version in the URI path.
+- Extend KafkaAgent to collect `kafka.log:type=LogManager,name=remainingLogsToRecover` and `kafka.log:type=LogManager,name=remainingSegmentsToRecover` metrics.
+- Use Jetty's native ServerHandler API to handle a `GET` request and return a JSON response that looks like this:
+```
+{
+  "brokerState": 2,
+  "remainingLogsToRecover": 123,
+  "remainingSegmentsToRecover": 456
+}
+```
+- Return HTTP 404 (Not Found) if a request is sent to an unknown version of the API e.g. `GET /v5/recovery-state` and `v5` does not exist.
+
+**KafkaRoller**:
+
+The proposed flow of KafkaRoller
+![The proposed flow of KafkaRoller](images/046-kafka-roller-new-flow.png)
+
+If a kafka pod does not become ready within the operational timeout (5 mins by default) take the following actions:
+1. Check if the maximum number of retries has been reached. If it has, fail the reconciliation.
+2. Ensure that the pod is not in a stuck state. The existing `isPodStuck()` can be used for this which returns true if the pod is in `CrashLoopBackOff`, `ImagePullBackOff`, `ContainerCreating` or `Pending` state. If the pod is stuck, continue with the current behaviour to retry rolling of the pod.
+3. Use the Pod DNS name to query the KafkaAgent API endpoint.
```

Comment:
wdym by "query" here? What is described at steps 5 and 6? So actually authenticating and sending the GET request? If yes, then 5 and 6 should be put here in the right order then moving 4 down because you can get a connection refused or timeout (as suggested by Jakub) only after you try to authenticate and submit the GET request, or?